toutoumomoma.go

```diff
@@ -328,31 +332,38 @@ func streamEntropy(r io.Reader) (float64, error) {
 		}
 		if err != nil {
 			if err != io.EOF {
-				return 0, err
+				return 0, 0, err
 			}
 			break
 		}
 	}
 
-	return entropy(&n, N), nil
+	h, sigma = entropy(&n, N)
+	return h, sigma, nil
 }
 
 // entropy returns the entropy for counts in n for a sequence that is N long.
-func entropy(n *[256]float64, N int) float64 {
+func entropy(n *[256]float64, N int) (h, sigma float64) {
+	if N == 0 {
+		return 0, 0
+	}
 	// e = -∑i=1..k((p_i)*log(p_i))
-	var e float64
-	f := 1 / float64(N)
+	scl := 1 / float64(N)
 	for _, cnt := range n {
 		if cnt == 0 {
 			// Ignore zero counts.
 			continue
 		}
-		p := f * cnt
-		e += p * math.Log2(p)
+		p := scl * cnt
+		l2p := math.Log2(p)
+		h += p * l2p
+		sigma += p * l2p * l2p
 	}
-	if e == 0 {
+	sigma -= h * h
```

Comment:
Possible cancellation error, leading to negative variance. Some better algorithms for calculating variance are here: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

Reply:
It's not entirely clear how to incorporate those into this calculation. Maybe compensated summation would be appropriate?

Reply:
I don't think this will help you, because you will still be calculating the difference of two large numbers in line 362 `sigma -= h * h`. 

You can do a two-pass algorithm, this will be more stable, but slower.

Reply:
The maximum of `h*h` here is 64 since we're calculating the entropy of a byte stream in bits. But maybe something the approach we use in stat [for the weighted case](https://github.com/gonum/gonum/blob/6b90107d4a29dcf67ae7c214a14935d0bcfbf129/stat/stat.go#L1340-L1348) would work?

Reply:
Has the same problem: `unnormalisedVariance = (ss - compensation*compensation/sumWeights)`

Reply:
I would try this: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Weighted_incremental_algorithm

(I suppose Gonum could also use it.)

Reply:
I've incorporated that now, and made a test against the naive implementation. They largely agree as you'll see from the tolerances in the test.

PTAL